ee/clickhouse/queries/property_optimizer.py

```diff
@@ -0,0 +1,112 @@
+from typing import Dict, List, Optional, Set, Tuple, Union, cast
+
+from rest_framework.exceptions import ValidationError
+
+from posthog.constants import PropertyOperatorType
+from posthog.models import Filter
+from posthog.models.entity import Entity
+from posthog.models.filters.mixins.utils import cached_property
+from posthog.models.filters.path_filter import PathFilter
+from posthog.models.filters.retention_filter import RetentionFilter
+from posthog.models.filters.stickiness_filter import StickinessFilter
+from posthog.models.property import Property, PropertyGroup
+
+
+class PropertyOptimizer:
```

Comment:
What is the relation between PropertyOptimizer and ColumnOptimizer. Do they have inter-dependencies, what is responsible for what?

Reply:
I don't _think_ they're interdependent, since `PropertyOptimizer` deals only with the "WHERE" clause, while `ColumnOptimizer` deals with the "SELECT" clause, and items in the "WHERE" clause, as long as they're using columns defined on the table, are independent of the select clause. 

Perhaps we should rename both to express this better?

I'm also happy to combine them together, but this seemed like a decently big conceptual difference to keep them separate.

ColumnOptimizer does indeed deal with properties, but with an end goal of figuring out what to use in the select query.

Reply:
It kind of feels like they're co-dependant logically even though SELECT and WHERE are separate. Maybe a mixin-based approach or having a `PropertyWhereClauseOptimizer` attribute `ColumnOptimizer` makes sense?

Reply:
Ah, I figured out the exact spot I was trying to communicate earlier. I think there is a dependency here between column optimizer, person query, and property optimizer. With the new filtering pattern, a simply person property pushdown won't always work as already discussed. However, we now have an extra case where certain person properties should be selected by the person subquery and used in the filtering in the outer query (to avoid costly DISTINCT_ID IN {subquery} filters). 

This means that we just need to keep clear the fields that the person query will retrieve/exclude based on the property filters. I added changes that reflect this. in _get_fields, we'll only omit the inner properties that are directly filtered and return the properties that need to be used in the outer query